Add LevelDependentRateTES node with state-of-charge dependent charge/discharge rates (#26)
EnergyModelsX/EnergyModelsHeat.jl: docs/src/nodes/leveldependentratetes.md

Reviewer: Please do not add a newline to fit it within ca 90 characters. It is a Markdown file where our standard is to use a new line for each new sentence. That makes maintenance significantly easier.

That must be adjusted.

In addition, you should follow the standard approach for describing a node. I do prefer that all information for a given node is on a single page.

    charge and discharge capacities become decision variables.
    The slopes and intercepts of the c-rate curves then depend on those capacity variables, so
    the rate limit would multiply a capacity variable by the storage-level variable.
    This product of two decision variables is a *bilinear* (nonconvex quadratic) term, which
    the linear and mixed-integer-linear solvers used in this package (*e.g.*, `HiGHS`) cannot
    handle.

!!! danger "StorageBehavior for LevelDependentRateTES"
    Like the other thermal energy storage nodes, a [`LevelDependentRateTES`](@ref) can only
    use [`Cyclic`](@extref EnergyModelsBase.Cyclic) storage behaviors, and when using
    `RepresentativePeriods` this is reduced to
    [`CyclicRepresentative`](@extref EnergyModelsBase.CyclicRepresentative).

## [Introduced type and its fields](@id nodes-LDRTES-fields)

[`LevelDependentRateTES`](@ref) uses the same standard fields as the other thermal energy
storage nodes (see *[ThermalEnergyStorage nodes](@ref nodes-TES-fields-stand)*): `id`,
`charge`, `level`, `discharge`, `stor_res`, `input`, `output` and `data`.

> In addition it introduces the heat loss factor and the c-rate anchor points.

Reviewer: Please write the values out. The meaning of charge and discharge has changed as you must use a capacity `AbstractStorageParameters`, *i.e.*, `UnionCapacity` for all of the 3 capacities. Otherwise, it would not work.

  (the full installed discharge rate at a full storage) and runs through the provided points in
  *strictly descending* level order.
  All rates must be non-negative, the charge levels must lie in ``(0, capacity(level(n))]`` and
  the discharge levels in ``[0, capacity(level(n)))``.
  The number of anchor points sets the number of linear segments:
    - one point gives a single straight line,
    - two or three points give a piecewise-linear curve, which introduces binary variables to
      select the active segment.

As an example, the storage in `examples/advanced_example.jl` has an installed charge rate of
30, discharge rate of 40 and level capacity of 100, combined with the three-point curves

```julia
c_rate_points_charge    = [[10.0, 30.0], [50.0, 20.0], [100.0, 10.0]]
c_rate_points_discharge = [[75.0, 40.0], [50.0, 20.0], [25.0, 15.0]]
```

The charge curve therefore stays at the full rate of 30 up to a level of 10 and then falls to
20 at level 50 and 10 at the full level of 100.
The discharge curve reaches its full rate of 40 for levels at or above 75 and falls to 20 at
level 50 and 15 at level 25 as the storage empties.
The resulting theoretical curves are shown below.

> Near a full storage the charge rate is additionally limited by the remaining free capacity, and
> near an empty storage the discharge rate is limited by the available energy, which is why both
> curves bend towards zero at the respective ends.

Reviewer: ```suggestion
The charge rate is additionally limited by the remaining free capacity near a full storage , and
near an empty storage the discharge rate is limited by the available energy, which is why both
curves bend towards zero at the respective ends.
```

Sounds better

![Theoretical charge and discharge c-rate curves for the example anchor points.](crate_curves_example.png)

!!! tip "Visualising the curves"

>     The plot above was produced with the function
>     [`visualize_c_rates`](@ref EnergyModelsHeat.visualize_c_rates), which plots the theoretical
>     (dis-)charge curves a given set of anchor points produces, so you can check your input
>     before solving a model.
>     It is provided through a package extension and becomes available once you run
>     `using Plots`.

Reviewer: ```suggestion
    The plot above was produced with the function [`visualize_c_rates`](@ref EnergyModelsHeat.visualize_c_rates), which plots the theoretical (dis-)charge curves for a given set of anchor points.
    This allows you to check your input before solving a model.
    It is provided through a package extension and becomes available once you run `using Plots`.
```

## [Mathematical description](@id nodes-LDRTES-math)

In the following description we use the variable and function names from the model, as in the
*[ThermalEnergyStorage](@ref nodes-TES-math)* page.

### [Variables](@id nodes-LDRTES-math-var)

[`LevelDependentRateTES`](@ref) uses all standard [`RefStorage`](@extref EnergyModelsBase.RefStorage)
variables (see *[ThermalEnergyStorage variables](@ref nodes-TES-math-var)*).
In addition, when a c-rate curve uses more than one anchor point, the following auxiliary
binary variables are declared (in [`EMB.variables_node`](@ref EnergyModelsBase.variables_node)):

- ``\texttt{bin\_region\_charge}[n, t, 1:2]`` selects the active segment of the charge curve.
- ``\texttt{bin\_region\_discharge}[n, t, 1:2]`` selects the active segment of the discharge curve.

> Two binaries per direction are sufficient to encode up to three segments.

Reviewer: Use the same approach as for the other nodes, i.e., wit the sections `Standard Variables` and `additional variables`.

I want all variables of a given node available at a single page. This results in repetitions, but simplifies the understanding.

> ### [Constraints](@id nodes-LDRTES-math-con)

Reviewer: Use the same format as for [`ThermalEnergyStorage`](https://energymodelsx.github.io/EnergyModelsHeat.jl/stable/nodes/thermalenergystorage/#nodes-TES)`.

You can skip the level constraints if you provide a link to the thermal energy storage page.
